src/lib/engine/utils.js

```diff
@@ -39,3 +39,11 @@ export function makeURLsMarkdown(text) {
 
   return transformedText
 }
+
+export const indicesOf = (list, predFn) => list.map((e, i) => predFn(e) ? i : null).filter(e => _.isFinite(e))
+export function getSubMatches(str, regex, matchIndex) {
```

Comment:
Documentation and/or testing would be quite nice for some of these.  At least, a small example or something.